packages/wast-parser/src/tokenizer.js

```diff
@@ -165,6 +175,144 @@ function tokenize(input: string) {
+   */
+  const regexToState = (
+    regex,
+    nextState,
+    n = 1,
+    allowSeparators = false
+  ) => () => {
+    if (allowSeparators) {
+      if (char === NUMERIC_SEPARATOR) {
+        if (regex.test(lookbehind())) {
+          // Consume the separator and stay in current state
+          return [allowSeparators, 1];
+        } else {
+          unexpectedCharacter();
+        }
+      }
+    }
+
+    if (regex.test(lookahead(n, 0))) {
+      return [nextState, n];
+    }
+
+    return false;
+  };
+
+  /**
+   * Combines a list of transition functions into a complete state transition function.
+   * When none of the given transitions apply, the given default state is returned.
+   *
+   * @param {Array} transitions A list of transition functions
+   * @param {String} defaultState The state to be returned in case no transition matches
+   *
+   * @return {Array} A two element array containing [newState, eatLength] where eatLength is the amount of characters to consume.
+   *
+   */
+  const combineTransitions = (transitions, defaultState) => () => {
+    let newState = defaultState;
+    let match = false;
+    for (let i = 0; i < transitions.length; ++i) {
+      match = transitions[i]();
+      if (match) {
+        break;
+      }
+    }
+
+    return match;
+  };
+
+  const START = "START";
+  const HEX = "HEX";
+  const HEX_FRAC = "HEX_FRAC";
+  const NAN_HEX = "NAN_HEX";
+  const DEC = "DEC";
+  const DEC_UNDERSCORE = "DEC_UNDERSCORE";
+  const DEC_FRAC = "DEC_FRAC";
+  const DEC_EXP = "DEC_EXP";
+  const DEC_SIGNED_EXP = "DEC_SIGNED_EXP";
+  const DEC_EXP_UNDERSCORE = "DEC_EXP_UNDERSCORE";
+  const STOP = "STOP";
+  const HEX_SIGNED_EXP = "HEX_SIGNED_EXP";
+  const HEX_EXP = "HEX_EXP";
+  const HEX_UNDERSCORE = "HEX_UNDERSCORE";
+  const HEX_FRAC_UNDERSCORE = "HEX_FRAC_UNDERSCORE";
+  const HEX_EXP_UNDERSCORE = "HEX_EXP_UNDERSCORE";
+
+  /**
+   * Build the FSM for number literals
+   */
+  const numberLiteralFSM = {
+    START: combineTransitions(
+      [
+        regexToState(/-|\+/, START),
+        regexToState(/nan:0x/, NAN_HEX, 6),
+        regexToState(/nan|inf/, STOP, 3),
+        regexToState(/0x/, HEX, 2),
+        regexToState(/[0-9]/, DEC),
+        regexToState(/\./, DEC_FRAC)
+      ],
+      STOP
+    ),
+    DEC_FRAC: combineTransitions(
+      [
+        regexToState(/[0-9]/, DEC_FRAC, 1, DEC_FRAC),
+        regexToState(/e|E/, DEC_SIGNED_EXP)
+      ],
+      STOP
+    ),
+    DEC: combineTransitions(
+      [
+        regexToState(/[0-9]/, DEC, 1, DEC),
+        regexToState(/\./, DEC_FRAC),
+        regexToState(/e|E/, DEC_SIGNED_EXP)
+      ],
+      STOP
+    ),
+    DEC_SIGNED_EXP: combineTransitions(
+      [regexToState(/\+|-/, DEC_EXP), regexToState(/[0-9]/, DEC_EXP)],
+      STOP
+    ),
+    DEC_EXP: combineTransitions([regexToState(/[0-9]/, DEC_EXP, 1, DEC_EXP)]),
+
+    HEX: combineTransitions(
+      [
+        regexToState(/[0-9|A-F|a-f]/, HEX, 1, HEX),
```

Comment:
isn't the fourth arg of regexToState a boolean?

Reply:
Ah forgot to update the docs there: The problem is that regexToState needs to stay in the current state when a separator is encountered, but it doesnt have access to the state variable here. So I pass the state instead of the boolean, but Id prefer this to be a boolean. Ill see if I can do that